File: BinarySearch2/problem3.py
#Time Complexity: O(logn)
#Space Complexity: O(1)
#Did this code successfully run on leetcode: yes
#Any problem you faced while coding thing:
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        #idea : if we always go up until all the elements around are smaller (no where to go up then), it's a peak
        #since it's any peak it doesn't matter if it's a valley, we just go up any side
        # it is binary search because we're getting rid of half of the search space every iteration
        #if it's greater on the right, move low to mid+1, vice-versa
        low=0
        high=len(nums)-1
        while high>=low:
            mid=int(low+(high-low)/2) #to avoid overflow
            if (mid==0 or nums[mid-1]<nums[mid]) and (mid==len(nums)-1 or nums[mid+1]<nums[mid]):
                return mid
            if (nums[mid+1]>=nums[mid]):
                low=mid+1
            else:
                high=mid-1

  #!! check lecture to make sure

File: binarySearch-2/src/Problem3.java
//TC : O(logn) as binary search is used
//SC : O(1)

// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
//
public class Problem3 {
        public int findPeakElement(int[] nums) {


            int mid = 0;
            int low = 0;
            int high = nums.length - 1;

            while (low < high)
            {
                mid = low + (high-low)/2;

                //Note [mid] means element at index equal to mid.
                //Tere are two cases that can happen :
                //case 1 : The element to the right of mid is lesser than [mid] . In this case, the peak is on the left half
                //of the array including mid. so, we set high = mid and continue search

                //case 2: The element to the right is greater than the [mid], in this case, the peak definitely lies in the right
                //half and does not include [mid] as it cant be a peak

                //if this is true it means that the peak lies in the left part including [mid], hence we set high=mid
                if (nums[mid] > nums[mid + 1])
                    high = mid;

                    //if this condition is true that means peak lies in the right part excluding [mid], hence we set low = mid-1
                else if (nums[mid] < nums[mid + 1])
                    low = mid + 1;
            }

            //while loop is exited when low == high, hence we can return either low or high
            return high;

        }

}
